### PR TITLE
Customisations v2

### PR DESCRIPTION
**Pull request by valeriadamante**
This PR is made in order to integrate the possibility to customise the configurations for running the anatuple production

### CODE REVIEW

File: config/samples_Run2_2018.yaml [hunk 1]
```diff
@@ -33,6 +33,10 @@ GLOBAL:
       VSe: VVLoose
       VSmu: VLoose
       VSjet: Medium
+  corrections:
+    - tauID
+  deepTauVersion:
+   2p5
```
Comment (kandrosov) [inline config/samples_Run2_2018.yaml:39]:
```suggestion
  deepTauVersion: 2p5
```

File: Common/Utilities.py [hunk 2]
```diff
@@ -75,4 +75,22 @@ def Apply(self, func, *args, **kwargs):
...
             return result[1:]
         else:
-            self.df = result
\ No newline at end of file
+            self.df = result
+
+def ApplyConfigCustomisations(config_dict, customisations):
+    customisations_to_apply = customisations.split(',')
+    for customisation in customisations_to_apply:
+        substrings = customisation.split('=')
```
Comment (kandrosov) [inline Common/Utilities.py:83]: check that there is only 2 substrings

File: Common/Utilities.py [hunk 3]
```diff
@@ -75,4 +75,22 @@ def Apply(self, func, *args, **kwargs):
...
+        value = substrings[-1]
+        key_entries = substrings[0].split('.')
+        cfg_entry = config_dict
+        for key in key_entries[:-1]:
+            cfg_entry = cfg_entry[key]
+        cfg_entry[key_entries[-1]] = value
```
Comment (kandrosov) [inline Common/Utilities.py:89]:
```suggestion
        entry_type = type(cfg_entry[key_entries[-1]])
        cfg_entry[key_entries[-1]] = entry_type(value)
```

File: Common/Utilities.py [hunk 4]
```diff
@@ -75,4 +75,22 @@ def Apply(self, func, *args, **kwargs):
...
+    print(config_dict)
```
Comment (kandrosov) [inline Common/Utilities.py:90]:
```suggestion
```

File: Common/Utilities.py [hunk 5]
```diff
@@ -75,4 +75,22 @@ def Apply(self, func, *args, **kwargs):
...
+
+    with open(args.anaCache, 'r') as f:
+        anaCache = yaml.safe_load(f)
+
+    if os.path.exists(args.outFile):
+        os.remove(args.outFile)
```
Comment (kandrosov) [inline Common/Utilities.py:96]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]
also implement customisation in law jobs

File: config/samples_Run2_2018.yaml (hunk 1 above)
Comment (valeriadamante) [reply config/samples_Run2_2018.yaml:40]: ok

File: Common/Utilities.py (hunk 3 above)
Comment (valeriadamante) [reply Common/Utilities.py:89]: ok

File: Common/Utilities.py (hunk 5 above)
Comment (valeriadamante) [reply Common/Utilities.py:96]: ok

File: Common/Utilities.py (hunk 2 above)
Comment (valeriadamante) [reply Common/Utilities.py:83]: yep

File: Common/Utilities.py (hunk 4 above)
Comment (valeriadamante) [reply Common/Utilities.py:90]: ok

File: .gitignore [hunk 6]
```diff
@@ -27,3 +27,5 @@ Studies/Visual
 /filesEntries.txt
 /filesEntries_SingleMuon2018A.txt
 /cmssw_report.xml
+Corrections/data/*/*/*.root
+*.txt
```
Comment (kandrosov) [inline .gitignore:31]: there are too many entries in `.gitignore` keep only 
```
/.law
/data*
__pycache__
/output
/soft
/crab
/.crabOverseer*
```

**Review by kandrosov — CHANGES_REQUESTED** [review]
please, also:
- add full sample selection
- add list of channels to be stored
- fix the following line in this PR (to not create PR because of one line)
https://github.com/cms-hh-bbtautau/Framework/blob/main/AnaProd/tasks.py#L145
should be `sorted(all_pre_branches.items())`

File: AnaProd/anaCacheProducer.py [hunk 7]
```diff
@@ -52,9 +53,12 @@ def createAnaCache(inDir, outFile, global_params, range=None, verbose=1):
...
     parser.add_argument('--outFile', required=True, type=str)
     parser.add_argument('--nEvents', type=int, default=None)
-
+    parser.add_argument('--customisations', type=str, default="")
     args = parser.parse_args()
 
     with open(args.config, 'r') as f:
         config = yaml.safe_load(f)
+
+    if(args.customisations!=""):
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaCacheProducer.py:62]: ok

File: run_tools/law_customizations.py [hunk 8]
```diff
@@ -75,6 +75,7 @@ def __init__(self, *args, **kwargs):
         super(Task, self).__init__(*args, **kwargs)
         self.cmssw_env_ = None
         self.sample_config = os.path.join(self.ana_path(), 'config', f'samples_{self.period}.yaml')
+        self.customisations = ""
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply run_tools/law_customizations.py:78]: ok

File: .gitignore (hunk 6 above)
Comment (valeriadamante) [reply .gitignore:31]: ok

File: config/samples_Run2_2018.yaml [hunk 9]
```diff
@@ -33,12 +33,52 @@ GLOBAL:
       VSe: VVLoose
       VSmu: VLoose
       VSjet: Medium
+  corrections:
+    - tauID
+    - tauES
+  deepTauVersion: 2p5
```
Comment (kandrosov) [inline config/samples_Run2_2018.yaml:39]: for the moment keep 2p1 as default value

File: run_tools/law_customizations.py [hunk 10]
```diff
@@ -76,7 +76,7 @@ def __init__(self, *args, **kwargs):
         self.cmssw_env_ = None
         self.sample_config = os.path.join(self.ana_path(), 'config', f'samples_{self.period}.yaml')
 
-    def load_sample_configs(self):
+    def load_sample_configs(self, customisations=""):
```
Comment (kandrosov) [inline run_tools/law_customizations.py:79]:
```suggestion
    def load_sample_configs(self):
```

**Review by kandrosov — CHANGES_REQUESTED** [review]
remove Common/mergingNtuple.py

File: config/samples_Run2_2018.yaml (hunk 9 above)
Comment (valeriadamante) [reply config/samples_Run2_2018.yaml]: ok

File: run_tools/law_customizations.py (hunk 10 above)
Comment (valeriadamante) [reply run_tools/law_customizations.py:79]: ok

File: AnaProd/tasks.py [hunk 11]
```diff
@@ -35,7 +35,7 @@ def run(self):
             inDir = os.path.join(self.central_nanoAOD_path(), sample_name)
             os.makedirs(os.path.dirname(self.output().path), exist_ok=True)
             sh_call(['python3', producer, '--config', self.sample_config, '--inDir', inDir,
-                    '--outFile', self.output().path], env=self.cmssw_env())
+                    '--outFile', self.output().path, '--customisations', load_customisations()], env=self.cmssw_env())
```
Comment (kandrosov) [inline AnaProd/tasks.py:38]: there is no such method. add customisation as luigi.Parameter and use it

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/anaTupleProducer.py [hunk 12]
```diff
@@ -50,6 +50,8 @@ def addAllVariables(dfw, syst_name, isData, trigger_class):
     dfw.Define(f"Muon_recoJetMatchIdx", f"FindMatching(Muon_p4, Jet_p4, 0.5)")
     dfw.Define( f"Electron_recoJetMatchIdx", f"FindMatching(Electron_p4, Jet_p4, 0.5)")
     dfw.DefineAndAppend("channelId","static_cast<int>(httCand.channel())")
+    channel_to_select = " || ".join(f"channelId==static_cast<int>(Channel::{ch})" for ch in config["GLOBAL"]["channelSelection"])
```
Comment (kandrosov) [inline AnaProd/anaTupleProducer.py:53]:
```suggestion
    channel_to_select = " || ".join(f"httCand.channel()==Channel::{ch}" for ch in config["GLOBAL"]["channelSelection"])
```

File: AnaProd/anaTupleProducer.py [hunk 13]
```diff
@@ -50,6 +50,8 @@ def addAllVariables(dfw, syst_name, isData, trigger_class):
...
+    dfw.Filter(channel_to_select)
```
Comment (kandrosov) [inline AnaProd/anaTupleProducer.py:54]:
```suggestion
    dfw.Filter(channel_to_select, "select channels")
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/anaTupleProducer.py (hunk 12 above)
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:53]: ok

File: AnaProd/anaTupleProducer.py (hunk 13 above)
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:54]: ok

File: AnaProd/tasks.py (hunk 11 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:38]: ok

**Review by kandrosov — APPROVED** [review]

**Pull request merged**